pkg/controllers/traffic/traffic_controller.go

```diff
@@ -19,46 +19,116 @@ package traffic
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 
-	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/_internal/metadata"
+	kuadrantv1 "github.com/Kuadrant/multi-cluster-traffic-controller/pkg/apis/v1"
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/dns"
 	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/traffic"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	trafficFinalizer = "kuadrant.io/traffic-managment"
 )
 
 // Reconciler reconciles a traffic object
 type Reconciler struct {
 	WorkloadClient client.Client
-	ControlClient  client.Client
+	Hosts          HostService
+	Certificates   CertificateService
+}
+
+type HostService interface {
+	EnsureManagedHost(ctx context.Context, t traffic.Interface) ([]string, []*kuadrantv1.DNSRecord, error)
+	AddEndPoints(ctx context.Context, t traffic.Interface) error
+	RemoveEndpoints(ctx context.Context, t traffic.Interface) error
+}
+
+type CertificateService interface {
+	EnsureCertificate(ctx context.Context, host string, owner metav1.Object) error
+	GetCertificateSecret(ctx context.Context, host string) (*v1.Secret, error)
 }
 
 func (r *Reconciler) Handle(ctx context.Context, o runtime.Object) (ctrl.Result, error) {
 	_ = log.FromContext(ctx)
-
 	trafficAccessor := o.(traffic.Interface)
 	log.Log.Info("got traffic object", "kind", trafficAccessor.GetKind(), "name", trafficAccessor.GetName(), "namespace", trafficAccessor.GetNamespace())
-
-	if trafficAccessor.GetAnnotations() == nil {
-		return ctrl.Result{}, fmt.Errorf("expected dummy configmap annotation missing")
+	controllerutil.AddFinalizer(trafficAccessor, trafficFinalizer)
+	// TODO add in deletion logic
+	if trafficAccessor.GetDeletionTimestamp() != nil && !trafficAccessor.GetDeletionTimestamp().IsZero() {
+		// targets, err := trafficAccessor.GetDNSTargets()
+		if err := r.Hosts.RemoveEndpoints(ctx, trafficAccessor); err != nil {
+			return ctrl.Result{}, err
+		}
+		controllerutil.RemoveFinalizer(trafficAccessor, trafficFinalizer)
+		return ctrl.Result{}, nil
 	}
 
-	cmName := trafficAccessor.GetAnnotations()["configmap"]
-	cmField := trafficAccessor.GetAnnotations()["field"]
-
-	cm := &corev1.ConfigMap{}
-
-	err := r.WorkloadClient.Get(ctx, client.ObjectKey{Namespace: trafficAccessor.GetNamespace(), Name: cmName}, cm)
-	if err != nil {
+	// verify host is correct
+	// no managed host assigned assign one
+	// create empty DNSRecord with assigned host
+	managedHosts, records, err := r.Hosts.EnsureManagedHost(ctx, trafficAccessor)
+	if err != nil && err != dns.AlreadyAssignedErr {
 		return ctrl.Result{}, err
 	}
-
-	value := cm.Data[cmField]
-
-	metadata.AddLabel(trafficAccessor, "dummy-configmap-value", value)
+	for i, managedHost := range managedHosts {
```

Comment:
I think this logic should be moved out to it's own function, it's quite verbose and would be better if the handle function were kept cleaner by just calling out to private functions and checking for errors.

My concern is that as the logic grows, this will become a very difficult function to scan through for the part you're interested in.

Reply:
could you take a look now? I think it is relatively simple. Also this logic is going to be spit soon between sync and webhook